src/code/z_camera.c

```diff
@@ -4367,101 +4370,136 @@ s32 Camera_Subj4(Camera* camera) {
-    Math3D_LineClosestToPoint(&rwData->unk_00, &sp6C.pos, eyeNext);
-    at->x = eyeNext->x + rwData->unk_00.dir.x;
-    at->y = eyeNext->y + rwData->unk_00.dir.y;
-    at->z = eyeNext->z + rwData->unk_00.dir.z;
+    Actor_GetWorldPosShapeRot(&playerPosRot, &camera->player->actor);
+    Math3D_LineClosestToPoint(&rwData->crawlspaceLine, &playerPosRot.pos, eyeNext);
+
+    // *at is unused before getting overwritten later this function
+    at->x = eyeNext->x + rwData->crawlspaceLine.dir.x;
+    at->y = eyeNext->y + rwData->crawlspaceLine.dir.y;
+    at->z = eyeNext->z + rwData->crawlspaceLine.dir.z;
+
     *eye = *eyeNext;
-    sp64.yaw = rwData->unk_30;
-    sp64.r = 5.0f;
-    sp64.pitch = 0x238C;
-    Camera_Vec3fVecSphGeoAdd(&sp98, eyeNext, &sp64);
-    rwData->unk_2C += 0xBB8;
-    temp_f16 = Math_CosS(rwData->unk_2C);
-    eye->x += (sp98.x - eye->x) * fabsf(temp_f16);
-    eye->y += (sp98.y - eye->y) * fabsf(temp_f16);
-    eye->z += (sp98.z - eye->z) * fabsf(temp_f16);
-
-    if ((rwData->unk_28 < temp_f16) && !rwData->unk_2E) {
+
+    targetOffset.yaw = rwData->yaw;
+    targetOffset.r = 5.0f;
+    targetOffset.pitch = 0x238C; // ~50 degrees
+
+    Camera_VecSphGeoAddToVec3f(&vEyeTarget, eyeNext, &targetOffset);
+
+    rwData->eyeLerpPhase += 0xBB8;
+    eyeLerp = Math_CosS(rwData->eyeLerpPhase);
+
+    // VEC3F_LERPIMPDST(eye, eye, &vEyeTarget, fabsf(eyeLerp))
+    eye->x += (vEyeTarget.x - eye->x) * fabsf(eyeLerp);
+    eye->y += (vEyeTarget.y - eye->y) * fabsf(eyeLerp);
+    eye->z += (vEyeTarget.z - eye->z) * fabsf(eyeLerp);
+
+    // When camera reaches the peak of offset and starts to move down
+    // && alternating cycles (sfx plays only every 2nd cycle)
+    if ((eyeLerp > rwData->eyeLerp) && !rwData->isSfxOff) {
         player = camera->player;
-        rwData->unk_2E = true;
-        func_800F4010(&player->actor.projectedPos, player->unk_89E + 0x8B0, 4.0f);
-    } else if (rwData->unk_28 > temp_f16) {
-        rwData->unk_2E = false;
+        rwData->isSfxOff = true;
+        func_800F4010(&player->actor.projectedPos, NA_SE_PL_CRAWL + player->unk_89E, 4.0f);
+    } else if (eyeLerp < rwData->eyeLerp) {
+        rwData->isSfxOff = false;
     }
 
-    rwData->unk_28 = temp_f16;
+    rwData->eyeLerp = eyeLerp;
+
     camera->player->actor.world.pos = *eyeNext;
     camera->player->actor.world.pos.y = camera->playerGroundY;
-    camera->player->actor.shape.rot.y = sp64.yaw;
-    temp_f16 = ((240.0f * temp_f16) * (rwData->unk_24 * 0.416667f));
-    temp_a0 = temp_f16 + rwData->unk_30;
-    at->x = eye->x + (Math_SinS(temp_a0) * 10.0f);
+    camera->player->actor.shape.rot.y = targetOffset.yaw;
+
+    eyeLerp = ((240.0f * eyeLerp) * (rwData->xzSpeed * 0.416667f));
+    xzOffsetTimer = eyeLerp + rwData->yaw;
+
+    at->x = eye->x + (Math_SinS(xzOffsetTimer) * 10.0f);
     at->y = eye->y;
-    at->z = eye->z + (Math_CosS(temp_a0) * 10.0f);
+    at->z = eye->z + (Math_CosS(xzOffsetTimer) * 10.0f);
```

Comment:
`xzOffsetTimer` is the yaw from eye to at

typically `xzOffsetTimer ~ rwData->yaw` which is the crawlspace orientation (forwards or backwards I'm not sure, btw `Camera_XZAngle` may have "from" and "to" swapped... I'd say forwards but not sure)

as for the 10.0f and whatever goes into `eyeLerp`, to me it's just values that Work Well :tm:

Reply:
btw `rwData->yaw` could be nice as `rwData->yawForwards` or `rwData->yawBackwards` depending on which it is?

Reply:
Nice, yeah that's clearer. I double checked and `rwData->yaw` is always facing towards the forward crawling direction, so I've also renamed it and added a comment